## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :x: [more...](#540ae58c-dadc-41f4-8175-a7aac4a160ec) |
| mint-gateway-nas.sh | :x: [more...](#a6c93d5f-28e2-4a99-bac3-7c063cee8a38) |
<a name="a6c93d5f-28e2-4a99-bac3-7c063cee8a38"></a>
### 8335-5c6e159/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.59:32686
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 3192ba5b0ee1:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 41 seconds
(4/14) Running aws-sdk-ruby tests ... done in 2 seconds
(5/14) Running awscli tests ... done in 1 minutes and 53 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 57 seconds
(8/14) Running minio-dotnet tests ... done in 30 seconds
(9/14) Running minio-go tests ... done in 1 minutes and 2 seconds
(10/14) Running minio-java tests ... done in 1 minutes and 40 seconds
(11/14) Running minio-js tests ... done in 33 seconds
(12/14) Running minio-py tests ... FAILED in 2 minutes and 43 seconds
{
  "duration": 64,
  "args": {
    "bucket_name": "minio-py-test-85197708-fe63-4e09-bae1-f5da4e45d28d",
    "response.status": 403,
    "response.reason": "Forbidden",
    "response.headers": "{\"_container\": {\"accept-ranges\": [\"Accept-Ranges\", \"bytes\"], \"content-length\": [\"Content-Length\", \"404\"], \"content-security-policy\": [\"Content-Security-Policy\", \"block-all-mixed-content\"], \"content-type\": [\"Content-Type\", \"application/xml\"], \"server\": [\"Server\", \"MinIO/DEVELOPMENT.2019-09-30T10-46-37Z\"], \"vary\": [\"Vary\", \"Origin\"], \"x-amz-request-id\": [\"X-Amz-Request-Id\", \"15C932CA531F773A\"], \"x-xss-protection\": [\"X-Xss-Protection\", \"1; mode=block\"], \"date\": [\"Date\", \"Mon, 30 Sep 2019 11:09:49 GMT\"]}}",
    "response._body": "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>AccessDenied</Code><Message>Request has expired</Message><Key>2dd56610-b6fa-4121-acc8-d44dc3f1cc3a</Key><BucketName>minio-py-test-85197708-fe63-4e09-bae1-f5da4e45d28d</BucketName><Resource>/minio-py-test-85197708-fe63-4e09-bae1-f5da4e45d28d/2dd56610-b6fa-4121-acc8-d44dc3f1cc3a</Resource><RequestId>15C932CA531F773A</RequestId><HostId></HostId></Error>",
    "presigned_get_object_url": "http://72.28.97.59:32686/minio-py-test-85197708-fe63-4e09-bae1-f5da4e45d28d/2dd56610-b6fa-4121-acc8-d44dc3f1cc3a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20190930%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190930T110942Z&X-Amz-Expires=5&X-Amz-SignedHeaders=host&X-Amz-Signature=7e698f3f8fb6f00d65292876ac6a95cf1136936c321f3727de1d7338f317b083",
    "object_name": "2dd56610-b6fa-4121-acc8-d44dc3f1cc3a"
  },
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1341, in test_presigned_get_object_expiry_5sec\n    object_name).get_exception()\nminio.error.AccessDenied: AccessDenied: message: Access Denied\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2036, in main\n    test_presigned_get_object_expiry_5sec(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1355, in test_presigned_get_object_expiry_5sec\n    raise Exception(err)\nException: AccessDenied: message: Access Denied\n",
  "name": "minio-py:test_presigned_get_object_expiry_5sec",
  "message": "AccessDenied: message: Access Denied",
  "status": "FAIL",
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)"
}

Executed 11 out of 14 tests successfully.
```
<a name="540ae58c-dadc-41f4-8175-a7aac4a160ec"></a>
### 8335-5c6e159/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:31910
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c7fc87815f2c:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 44 seconds
(4/14) Running aws-sdk-ruby tests ... done in 5 seconds
(5/14) Running awscli tests ... done in 1 minutes and 7 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 34 seconds
(8/14) Running minio-dotnet tests ... done in 36 seconds
(9/14) Running minio-go tests ... done in 1 minutes and 48 seconds
(10/14) Running minio-java tests ... done in 2 minutes and 12 seconds
(11/14) Running minio-js tests ... done in 50 seconds
(12/14) Running minio-py tests ... FAILED in 3 minutes and 58 seconds
{
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)",
  "status": "FAIL",
  "name": "minio-py:test_presigned_get_object_expiry_5sec",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1341, in test_presigned_get_object_expiry_5sec\n    object_name).get_exception()\nminio.error.AccessDenied: AccessDenied: message: Access Denied\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2036, in main\n    test_presigned_get_object_expiry_5sec(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1355, in test_presigned_get_object_expiry_5sec\n    raise Exception(err)\nException: AccessDenied: message: Access Denied\n",
  "args": {
    "bucket_name": "minio-py-test-3ab84a26-fe29-4cb6-b344-6b25bf95dadc",
    "object_name": "77dee2c3-6497-40e0-bdb2-4a2fa4484025",
    "response.status": 403,
    "response.headers": "{\"_container\": {\"accept-ranges\": [\"Accept-Ranges\", \"bytes\"], \"content-length\": [\"Content-Length\", \"440\"], \"content-security-policy\": [\"Content-Security-Policy\", \"block-all-mixed-content\"], \"content-type\": [\"Content-Type\", \"application/xml\"], \"server\": [\"Server\", \"MinIO/DEVELOPMENT.2019-09-30T10-46-37Z\"], \"vary\": [\"Vary\", \"Origin\"], \"x-amz-request-id\": [\"X-Amz-Request-Id\", \"15C932E4DDCD4BD0\"], \"x-xss-protection\": [\"X-Xss-Protection\", \"1; mode=block\"], \"date\": [\"Date\", \"Mon, 30 Sep 2019 11:11:43 GMT\"]}}",
    "response._body": "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>AccessDenied</Code><Message>Request has expired</Message><Key>77dee2c3-6497-40e0-bdb2-4a2fa4484025</Key><BucketName>minio-py-test-3ab84a26-fe29-4cb6-b344-6b25bf95dadc</BucketName><Resource>/minio-py-test-3ab84a26-fe29-4cb6-b344-6b25bf95dadc/77dee2c3-6497-40e0-bdb2-4a2fa4484025</Resource><RequestId>15C932E4DDCD4BD0</RequestId><HostId>8a410fca-35f4-4493-a53d-d304a8e2281d</HostId></Error>",
    "response.reason": "Forbidden",
    "presigned_get_object_url": "http://72.28.97.51:31910/minio-py-test-3ab84a26-fe29-4cb6-b344-6b25bf95dadc/77dee2c3-6497-40e0-bdb2-4a2fa4484025?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20190930%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190930T111136Z&X-Amz-Expires=5&X-Amz-SignedHeaders=host&X-Amz-Signature=c728dffb3f5ae641a782be8df298dc11da6efd3987ba009f2f8c843560c7209f"
  },
  "message": "AccessDenied: message: Access Denied",
  "duration": 180
}

Executed 11 out of 14 tests successfully.
```

--- a/cmd/signature-v4-parser.go
+++ b/cmd/signature-v4-parser.go
@@ -250,6 +250,8 @@ func parsePreSignV4(query url.Values, region string, stype serviceType) (psv pre
 //            SignedHeaders=signedHeaders, Signature=signature
 //
 func parseSignV4(v4Auth string, region string, stype serviceType) (sv signValues, aec APIErrorCode) {
+	// credElement is fetched first to skip replacing the space in access key.
+	credElement := strings.TrimPrefix(strings.Split(strings.TrimSpace(v4Auth), ",")[0], signV4Algorithm)
 	// Replace all spaced strings, some clients can send spaced
 	// parameters and some won't. So we pro-actively remove any spaces
 	// to make parsing easier.
@@ -275,7 +277,7 @@ func parseSignV4(v4Auth string, region string, stype serviceType) (sv signValues
 
 	var err APIErrorCode
 	// Save credentail values.
-	signV4Values.Credential, err = parseCredentialHeader(authFields[0], region, stype)
+	signV4Values.Credential, err = parseCredentialHeader(strings.TrimSpace(credElement), region, stype)
 	if err != ErrNone {
 		return sv, err
 	}

--- a/cmd/signature-v4-parser_test.go
+++ b/cmd/signature-v4-parser_test.go
@@ -462,6 +462,36 @@ func TestParseSignV4(t *testing.T) {
 			},
 			expectedErrCode: ErrNone,
 		},
+		// Test case - 8.
+		{
+			inputV4AuthStr: signV4Algorithm +
+				strings.Join([]string{
+					// generating a valid credential.
+					generateCredentialStr(
+						"access key",
+						sampleTimeStr,
+						"us-west-1",
+						"s3",
+						"aws4_request"),
+					// valid SignedHeader.
+					"SignedHeaders=host;x-amz-content-sha256;x-amz-date",
+					// valid Signature field.
+					// a valid signature is of form "Signature="
+					"Signature=abcd",
+				}, ","),
+			expectedAuthField: signValues{
+				Credential: generateCredentials(
+					t,
+					"access key",
+					sampleTimeStr,
+					"us-west-1",
+					"s3",
+					"aws4_request"),
+				SignedHeaders: []string{"host", "x-amz-content-sha256", "x-amz-date"},
+				Signature:     "abcd",
+			},
+			expectedErrCode: ErrNone,
+		},
 	}
 
 	for i, testCase := range testCases {

